helm/prometheus-meta-operator/templates/prometheus-rules/network.all.rules.yml

```diff
@@ -75,3 +75,43 @@ spec:
         team: rocket
         {{- end }}
         topic: network
+    - alert: HighNumberOfAllocatedSockets
+      annotations:
+        description: '{{`Over 90 percent of sockets are allocated on {{ $labels.instance }}.`}}'
+        opsrecipe: high-port-usage/
+      # This threshold value is based off net.ipv4.ip_local_port_range = 32768 60999
+      expr: node_sockstat_TCP_alloc > ( 0.9 * ( 60999 - 32768 ) )
+      for: 5m
+      labels:
+        area: kaas
+        cancel_if_outside_working_hours: "true"
+        severity: page
+        team: ludacris
+        topic: network
+    - alert: HighNumberOfOrphanedSockets
+      annotations:
+        description: '{{`Over 80 percent of ports are orphaned {{  $labels.instance }}.`}}'
+        opsrecipe: high-port-usage/
+      # This threshold value is based off net.ipv4.ip_local_port_range = 32768 60999
+      expr: node_sockstat_TCP_orphan > ( 0.8 * ( 60999 - 32768 ) )
+      for: 5m
+      labels:
+        area: kaas
+        cancel_if_outside_working_hours: "true"
+        severity: page
+        team: ludacris
+        topic: network
+    - alert: HighNumberOfTimeWaitSockets
+      annotations:
+        description: '{{`Over 80 percent of available ports are in time wait state on {{ $labels.instance }}.`}}'
```

Comment:
Okay here is my question: Why alert if 80% of ports are in wait but 90% allocated? What's the difference between wait and allocated?

Like will it not anyway hit 90% allocated if 80% are in wait?

Reply:
> Like will it not anyway hit 90% allocated if 80% are in wait?

I think not necessarily as it would depend on workload behavior? Every workload could `FIN` at once and hit 80% tw without having 90% allocated overall, IIUC.

But my feedback might actually be the opposite of Marcel's -- why 80% here? Might that be too high to actually catch cases where the workload is "leaking" sockets? Or maybe we just care about extreme cases? I'm fine either way as this is exploratory, just curious if there's a reason for 80.

Reply:
Ports go into a `Time Wait (tw)` state when they are closed from the server side . Before the port can be reallocated, the kernel needs to wait for any client packets that might still be in transit or make sure it receives an ACK that the connection is closed on the client side. This is to prevent the packet from being routed to a new socket that might reuse the same port. If this count gets really high it usually indicates there are things that can be done to correct it. 

80% threshold is too high, I simply picked an arbitrary conservative number that is lower than 90%. It does makes more sense to lower it, I'll change it to 50%. If 50% of the total allocatable ports are stuck waiting and not allowed to be used, there are things that can be done to improve the socket handling.